packages/component-base/src/data-provider-controller/cache.js

```diff
@@ -0,0 +1,187 @@
+export class Cache {
+  /**
+   * @type {import('../data-provider-controller.js').DataProviderController}
+   */
+  controller;
+
+  /**
+   * The number of items.
+   *
+   * @type {number}
+   */
+  size = 0;
+
+  /**
+   * The number of items to display per page.
+   *
+   * @type {number}
+   */
+  pageSize;
+
+  /**
+   * The total number of items, including items from expanded sub-caches.
+   *
+   * @type {number}
+   */
+  effectiveSize = 0;
+
+  /**
+   * An array of cached items.
+   *
+   * @type {object[]}
+   */
+  items = [];
+
+  /**
+   * A map where the key is a requested page and the value is a callback
+   * that will be called with data once the request is complete.
+   *
+   * @type {Map<number, Function>}
+   */
+  pendingRequests = new Map();
+
+  /**
+   * A map where the key is the index of an item in the `items` array
+   * and the value is a sub-cache associated with that item.
+   *
+   * Note, it's intentionally defined as an object instead of a Map
+   * to ensure that Object.entries() returns an array with keys sorted
+   * in alphabetical order, rather than the order they were added.
+   *
+   * @type {Record<number, Cache>}
+   */
+  #subCacheByIndex = {};
+
+  /**
+   * @param {Cache['controller']} controller
+   * @param {number} pageSize
+   * @param {number | undefined} size
+   * @param {Cache | undefined} parentCache
+   * @param {number | undefined} parentCacheIndex
+   */
+  constructor(controller, pageSize, size, parentCache, parentCacheIndex) {
+    this.controller = controller;
+    this.pageSize = pageSize;
+    this.size = size || 0;
+    this.effectiveSize = size || 0;
+    this.parentCache = parentCache;
+    this.parentCacheIndex = parentCacheIndex;
+  }
+
+  /**
+   * An item in the parent cache that the current cache is associated with.
+   *
+   * @return {object | undefined}
+   */
+  get parentItem() {
+    return this.parentCache && this.parentCache.items[this.parentCacheIndex];
+  }
+
+  /**
+   * An array of sub-caches sorted in the same order as their associated items
+   * appear in the `items` array.
+   *
+   * @return {Cache[]}
+   */
+  get subCaches() {
+    return Object.values(this.#subCacheByIndex);
+  }
+
+  /**
+   * Whether the cache or any of its descendant caches have pending requests.
+   *
+   * @return {boolean}
+   */
+  get isLoading() {
+    if (this.pendingRequests.size > 0) {
+      return true;
+    }
+
+    return this.subCaches.some((subCache) => subCache.isLoading);
+  }
+
+  /**
+   * Recalculates the effective size for the cache and its descendant caches recursively.
+   */
+  recalculateEffectiveSize() {
+    this.effectiveSize =
+      !this.parentItem || this.controller.isExpanded(this.parentItem)
```

Comment:
I kind of liked the revision where `cache` didn't know about the controller but only had a reference to an `isExpanded` function. In the latest revision, these two modules are co-dependent. Maybe the constructor could accept any object that includes an `isExpanded` function?

Reply:
Good point. The controller instance is no longer required in the cache class. How about we pass a `context` object instead?

```js
get cacheContext() {
  return { isExpanded: this.isExpanded } 
}

#createRootCache() {
  return new Cache(this.cacheContext, ...);
}
```

Reply:
I guess from the cache's point of view it wouldn't matter if it's the controller or a new object referencing a method in the controller. The cache only sees is as an object that has `isExpanded` method

Reply:
I figured that having a distinct `cacheContext` object would probably be better because it's more explicit.